iris_ugrid/tests/synthetic_data_generator/__init__.py

```diff
@@ -0,0 +1,115 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+"""
+Provides synthetic data generation capability in order to test ugrid loading and regridding.
+"""
+from pathlib import Path
+from string import Template
+import subprocess
+
+import netCDF4
+import numpy as np
+
+
+def create_file(
+    temp_file_dir,
+    dataset_name,
+    dataset_type,
+    n_faces=866,
+    n_times=1,
+    n_levels=39,
+):
+    """Create a synthetic NetCDF file with XIOS-like content.
+
+    Starts from a template CDL headers string, modifies to the input dimensions
+    then adds data of the correct size.
+
+    Parameters
+    ----------
+    temp_file_dir : str or pathlib.Path
+        The directory in which to place the created file.
+    dataset_name : str
+        The name for the NetCDF dataset and also the created file.
+    dataset_type : str
+        Name of the type of dataset to write to the file. The name = the name
+        of the template CDL string file to fetch from the `mesh_headers/`
+        directory.
+    n_faces, n_times, n_levels : int
+        Descriptors for the desired dimensions of the file's dataset.
+
+    Returns
+    -------
+    str
+        Path of the created NetCDF file.
+
+    """
+
+    # Fetch the specified CDL template type, modify the placeholders for the
+    # specified dimensions.
+    templates_dir = Path(__file__).parent / "mesh_headers"
+    template_filepath = templates_dir.joinpath(dataset_type).with_suffix(
+        ".txt"
+    )
+    with open(template_filepath) as file:
+        template_string = Template(file.read())
+    cdl = template_string.substitute(
+        DATASET_NAME=dataset_name,
+        NUM_NODES=n_faces + 2,
+        NUM_FACES=n_faces,
+        NUM_LEVELS=n_levels,
+    )
+
+    # Spawn an "ncgen" command to create an actual netcdf file from the cdl string.
+    nc_filepath = Path(temp_file_dir).joinpath(dataset_name).with_suffix(".nc")
+    subprocess.run(
+        ["ncgen", "-o" + str(nc_filepath)],
+        input=cdl,
+        encoding="ascii",
+        check=True,
+    )
+
+    # Load the netcdf file and post-modify it to define the variable data contents.
+    ds = netCDF4.Dataset(nc_filepath, "r+")
+
+    def should_be_monotonic(variable):
```

Comment:
I think this will apply the fix to too many variables, but we only really needed to "fix" dimension variables.
 -- so the test for that can be something like 
`len(variable.dimensions) == 1 and variable.dimensions[0] == variable.name`.
( But ... can you also _check that still works_ , as the measure of that is currently somewhere else : no tests here for what becomes a dim coord ?  Sorry for the fuss )


In any case, I really don't like the special detection of `'Mesh'` and `'nMesh'` in names, because that only works for XIOS-like datafiles..

Reply:
4559730c3a769f292d82c87c887ddd88f2704bc7